src/lib/zexe_backend/zexe_backend_common/plonk_plookup_constraint_system.ml

```diff
@@ -0,0 +1,1172 @@
+            Poseidon prev.(0) prev.(1) prev.(2) prev.(3) prev.(4)
+            Params.params.round_constants.(ind)
+        in
+        Array.iteri
+          ~f:(fun i perm ->
+            if i = Array.length state - 1 then
+              let prev =
+                Array.mapi perm ~f:(fun i x ->
+                    wire sys x sys.next_row (index_to_col i) )
+              in
+              add_row sys
+                (Array.map perm ~f:(fun x -> Some x))
+                Zero prev.(0) prev.(1) prev.(2) prev.(3) prev.(4) [||]
+            else add_round_state perm i )
+          state
+    | Plonk_constraint.T (EC_add {p1; p2; p3; r}) ->
+        let row =
+          Array.map
+            [|fst p1; snd p1; fst p2; snd p2; r|]
+            ~f:(fun x -> reduce_to_v x)
+        in
+        add_row sys
+          (Array.map row ~f:(fun x -> Some x))
+          Add
+          (wire sys row.(0) sys.next_row L)
+          (wire sys row.(1) sys.next_row R)
+          (wire sys row.(2) sys.next_row O)
+          (wire sys row.(3) sys.next_row Q)
+          (wire sys row.(4) sys.next_row P)
+          [||] ;
+        let x3 = reduce_to_v (fst p3) in
+        let y3 = reduce_to_v (snd p3) in
+        add_row sys
+          [|Some x3; Some y3; None; None; None|]
+          Zero
+          (wire sys x3 sys.next_row L)
+          (wire sys y3 sys.next_row R)
+          {row= After_public_input sys.next_row; col= O}
+          {row= After_public_input sys.next_row; col= Q}
+          {row= After_public_input sys.next_row; col= P}
+          [||] ;
+        ()
+    | Plonk_constraint.T (EC_double {p1; p2; r}) ->
+        let row =
+          Array.map
+            [|fst p1; snd p1; fst p2; snd p2; r|]
+            ~f:(fun x -> reduce_to_v x)
+        in
+        add_row sys
+          (Array.map row ~f:(fun x -> Some x))
+          Double
+          (wire sys row.(0) sys.next_row L)
+          (wire sys row.(1) sys.next_row R)
+          (wire sys row.(2) sys.next_row O)
+          (wire sys row.(3) sys.next_row Q)
+          (wire sys row.(4) sys.next_row P)
+          [||] ;
+        ()
+    | Plonk_constraint.T (EC_scale {state}) ->
+        let add_ecscale_round (round : V.t Scale_round_5_wires.t) =
+          let xt = wire sys round.xt sys.next_row L in
+          let yt = wire sys round.yt sys.next_row R in
+          let l1 = wire sys round.l1 sys.next_row O in
+          let l2 = wire sys round.l2 sys.next_row Q in
+          let b = wire sys round.b sys.next_row P in
+          let xs = wire sys round.xs (sys.next_row + 1) L in
+          let ys = wire sys round.ys (sys.next_row + 1) R in
+          let xp = wire sys round.xp (sys.next_row + 1) O in
+          let yp = wire sys round.yp (sys.next_row + 1) Q in
+          add_row sys
+            [| Some round.xt
+             ; Some round.yt
+             ; Some round.l1
+             ; Some round.l2
+             ; Some round.b |]
+            Vbmul1 xt yt l1 l2 b [||] ;
```

Comment:
It might be clearer if this was called `Vbmul` (and `Vbmul2` became `Vbmulpack` or similar)